File: mcp_servers/src/core/optimized_base_server.py
"""
Optimized MCP Server Base Class - Python Implementation
High-performance implementation with all optimization features enabled
"""

import asyncio
import logging
import time
import multiprocessing
import concurrent.futures
from typing import Any, Dict, List, Optional, Callable, Union
from dataclasses import dataclass, field
from abc import ABC, abstractmethod

from .base_server import BaseMCPServer, MCPServerOptions, HealthReport
from .performance_optimizer import PerformanceOptimizer, PerformanceConfig, PerformanceMetrics

@dataclass
class OptimizedServerConfig(MCPServerOptions):
    """Extended configuration for optimized server"""
    performance: Optional[PerformanceConfig] = None
    database: Optional[Dict[str, Any]] = None
    monitoring: Optional[Dict[str, Any]] = None
    clustering: Optional[Dict[str, Any]] = None

class OptimizedMCPServer(BaseMCPServer):
    """
    High-performance MCP server with all optimizations enabled:
    - Async/await patterns throughout
    - Worker thread/process pools for CPU-intensive tasks
    - Advanced caching (LRU + Redis)
    - Connection pooling
    - Performance monitoring
    - Memory optimization
    """
    
    def __init__(self, config: OptimizedServerConfig):
        super().__init__(config)
        
        # Initialize performance optimizer
        perf_config = config.performance or PerformanceConfig()
        self.performance_optimizer = PerformanceOptimizer(perf_config, self.logger)
        
        # Connection pools
        self.db_pool: Optional[Any] = None
        self.http_session: Optional[Any] = None
        
        # Monitoring
        self.metrics_task: Optional[asyncio.Task] = None
        
        # Setup optimizations
        asyncio.create_task(self._setup_optimizations())
    
    async def _setup_optimizations(self):
        """Setup all optimization features"""
        try:
            await self._setup_connection_pools()
            await self._setup_monitoring()
            self.logger.info("Performance optimizations initialized")
        except Exception as e:
            self.logger.error(f"Failed to setup optimizations: {e}")
    
    async def _setup_connection_pools(self):
        """Setup database and HTTP connection pools"""
        # Database connection pool setup would go here
        # This is a placeholder for actual database pool implementation
        pass
    
    async def _setup_monitoring(self):
        """Setup performance monitoring"""
        if self.performance_optimizer.config.enable_metrics:
            self.metrics_task = asyncio.create_task(self._metrics_collection_loop())
    
    async def _metrics_collection_loop(self):
        """Continuous metrics collection"""
        while True:
            try:
                await asyncio.sleep(self.performance_optimizer.config.metrics_interval)
                # Metrics are collected automatically by the performance optimizer
            except asyncio.CancelledError:
                break
            except Exception as e:
                self.logger.error(f"Metrics collection error: {e}")
                await asyncio.sleep(5)
    
    # Enhanced request handling with performance optimizations
    async def handle_optimized_request(
        self,
        request_name: str,
        handler: Callable[[], Any],
        cache_key: Optional[str] = None,
        cache_ttl: Optional[int] = None
    ) -> Any:
        """Handle request with performance optimizations"""
        start_time = time.time()
        self.performance_optimizer.track_request()
        
        try:
            # Try cache first if cache key provided
            if cache_key:
                cached_result = await self.performance_optimizer.get(cache_key)
                if cached_result is not None:
                    response_time = (time.time() - start_time) * 1000
                    self.performance_optimizer.track_response_time(response_time)
                    return cached_result
            
            # Process request asynchronously
            if asyncio.iscoroutinefunction(handler):
                result = await self.performance_optimizer.process_async(handler)
            else:
                result = await self.performance_optimizer.process_in_thread(handler)
            
            # Cache result if cache key provided
            if cache_key:
                await self.performance_optimizer.set(cache_key, result, cache_ttl)
            
            response_time = (time.time() - start_time) * 1000
            self.performance_optimizer.track_response_time(response_time)
            
            self.logger.info(
                f"Request {request_name} processed in {response_time:.2f}ms"
            )
            
            return result
            
        except Exception as error:
            self.performance_optimizer.track_error()
            response_time = (time.time() - start_time) * 1000
            self.performance_optimizer.track_response_time(response_time)
            
            self.logger.error(
                f"Request {request_name} failed after {response_time:.2f}ms: {error}"
            )
            raise
    
    # CPU-intensive task processing
    async def process_cpu_intensive_task(self, func: Callable, *args, **kwargs) -> Any:
        """Process CPU-intensive task in process pool"""
        return await self.performance_optimizer.process_in_process(func, *args, **kwargs)
    
    async def process_io_intensive_task(self, func: Callable, *args, **kwargs) -> Any:
        """Process I/O-intensive task in thread pool"""
        return await self.performance_optimizer.process_in_thread(func, *args, **kwargs)
    
    # Database operations with connection pooling
    async def query_database(self, sql: str, params: Optional[List[Any]] = None) -> Any:
        """Execute database query with connection pooling"""
        if not self.db_pool:
            raise RuntimeError("Database pool not configured")
        
        # This would use the actual database pool
        # Placeholder implementation
        await asyncio.sleep(0.001)  # Simulate database query
        return {"sql": sql, "params": params, "result": "mock_result"}
    
    async def transaction_database(self, callback: Callable) -> Any:
        """Execute database transaction"""
        if not self.db_pool:
            raise RuntimeError("Database pool not configured")
        
        # This would implement actual transaction logic
        # Placeholder implementation
        return await callback()
    
    # Cache operations
    async def get_cached(self, key: str) -> Optional[Any]:
        """Get value from cache"""
        return await self.performance_optimizer.get(key)
    
    async def set_cached(self, key: str, value: Any, ttl: Optional[int] = None) -> None:
        """Set value in cache"""
        await self.performance_optimizer.set(key, value, ttl)
    
    async def invalidate_cache(self, key: str) -> None:
        """Invalidate cache entry"""
        await self.performance_optimizer.delete(key)
    
    async def clear_cache(self) -> None:
        """Clear all cache"""
        await self.performance_optimizer.clear()
    
    # Performance metrics
    async def get_performance_metrics(self) -> PerformanceMetrics:
        """Get current performance metrics"""
        return await self.performance_optimizer.get_metrics()
    
    async def get_detailed_health(self) -> Dict[str, Any]:
        """Get detailed health status including performance metrics"""
        base_health = await self.get_health()
        performance_metrics = await self.get_performance_metrics()
        
        return {
            "base_health": base_health.__dict__,
            "performance": {
                "cpu": performance_metrics.cpu,
                "memory": performance_metrics.memory,
                "cache": performance_metrics.cache,
                "requests": performance_metrics.requests,
                "connections": performance_metrics.connections
            },
            "optimizations": {
                "async_processing": True,
                "caching": "redis+lru",
                "connection_pooling": self.db_pool is not None,
                "monitoring": self.metrics_task is not None and not self.metrics_task.done()
            },
            "uptime": time.time() - self.start_time
        }
    
    # Enhanced tool execution with optimizations
    async def execute_tool_optimized(
        self, 
        name: str, 
        arguments: Dict[str, Any],
        use_cache: bool = True,
        cache_ttl: Optional[int] = None
    ) -> Any:
        """Execute tool with optimization features"""
        cache_key = None
        if use_cache:
            # Create cache key from tool name and arguments
            cache_key = f"tool:{name}:{hash(str(sorted(arguments.items())))}"
        
        return await self.handle_optimized_request(
            f"tool:{name}",
            lambda: self.execute_tool(name, arguments),
            cache_key,
            cache_ttl
        )
    
    # Enhanced resource reading with optimizations
    async def read_resource_optimized(
        self,
        uri: str,
        use_cache: bool = True,
        cache_ttl: Optional[int] = None
    ) -> Dict[str, Any]:
        """Read resource with optimization features"""
        cache_key = None
        if use_cache:
            cache_key = f"resource:{uri}"
        
        return await self.handle_optimized_request(
            f"resource:{uri}",
            lambda: self.read_resource_content(uri),
            cache_key,
            cache_ttl
        )
    
    # Batch processing utilities
    async def process_batch(
        self, 
        items: List[Any], 
        processor: Callable[[Any], Any],
        batch_size: int = 100,
        use_processes: bool = False
    ) -> List[Any]:
        """Process items in batches for better performance"""
        results = []
        
        for i in range(0, len(items), batch_size):
            batch = items[i:i + batch_size]
            
            if use_processes:
                # Process batch in parallel using process pool
                batch_results = await asyncio.gather(*[
                    self.performance_optimizer.process_in_process(processor, item)
                    for item in batch
                ])
            else:
                # Process batch in parallel using thread pool
                batch_results = await asyncio.gather(*[
                    self.performance_optimizer.process_in_thread(processor, item)
                    for item in batch
                ])
            
            results.extend(batch_results)
        
        return results
    
    # Cleanup and shutdown
    async def cleanup(self) -> None:
        """Enhanced cleanup with optimization cleanup"""
        self.logger.info("Starting optimized server cleanup")
        
        try:
            # Cancel metrics collection
            if self.metrics_task and not self.metrics_task.done():
                self.metrics_task.cancel()
                try:
                    await self.metrics_task
                except asyncio.CancelledError:
                    pass
            
            # Cleanup performance optimizer
            await self.performance_optimizer.cleanup()
            
            # Close database pool
            if self.db_pool:
                # This would close the actual database pool
                pass
            
            # Close HTTP session
            if self.http_session:
                # This would close the actual HTTP session
                pass
            
            self.logger.info("Optimized server cleanup completed")
            
        except Exception as e:
            self.logger.error(f"Error during cleanup: {e}")
        
        # Call parent cleanup
        await super().cleanup()

# Factory function for creating optimized servers
def create_optimized_server(
    server_class: type,
    config: OptimizedServerConfig
) -> OptimizedMCPServer:
    """Create an optimized MCP server instance"""
    
    # Apply default optimizations for AMD Ryzen 7 7800X3D
    if config.performance is None:
        config.performance = PerformanceConfig(
            enable_async_processing=True,
            thread_pool_size=16,  # Match CPU cores
            process_pool_size=8,  # Half the cores for CPU-intensive tasks
            enable_async_batching=True,
            max_memory_usage_mb=28 * 1024,  # 28GB for 32GB system
            gc_threshold=75,
            enable_memory_profiling=True,
            lru_cache_size=10000,
            cache_ttl=3600,
            enable_redis_cache=True,
            redis_url="redis://localhost:6379",
            db_connection_pool_size=32,
            http_connection_pool_size=64,
            enable_metrics=True,
            metrics_interval=1,
            alert_thresholds={
                'cpu_usage': 85.0,
                'memory_usage': 90.0,
                'response_time': 5000.0
            }
        )
    
    return server_class(config)

# Multi-process server management
class ClusteredServerManager:
    """Manager for running multiple server processes"""
    
    def __init__(
        self,
        server_class: type,
        config: OptimizedServerConfig,
        num_workers: Optional[int] = None
    ):
        self.server_class = server_class
        self.config = config
        self.num_workers = num_workers or multiprocessing.cpu_count()
        self.workers: List[multiprocessing.Process] = []
        self.logger = logging.getLogger(f"{config.name}-cluster")
    
    def start(self):
        """Start clustered servers"""
        self.logger.info(f"Starting {self.num_workers} worker processes")
        
        for i in range(self.num_workers):
            worker = multiprocessing.Process(
                target=self._run_worker,
                args=(i,),
                name=f"{self.config.name}-worker-{i}"
            )
            worker.start()
            self.workers.append(worker)
        
        # Wait for workers
        try:
            for worker in self.workers:
                worker.join()
        except KeyboardInterrupt:
            self.logger.info("Shutting down cluster")
            self.stop()
    
    def stop(self):
        """Stop all workers"""
        for worker in self.workers:
            if worker.is_alive():
                worker.terminate()
                worker.join(timeout=30)
                if worker.is_alive():
                    worker.kill()
        
        self.workers.clear()
        self.logger.info("Cluster stopped")
    
    def _run_worker(self, worker_id: int):
        """Run a single worker process"""
        try:
            # Create worker-specific config
            worker_config = OptimizedServerConfig(
                name=f"{self.config.name}-worker-{worker_id}",
                version=self.config.version,
                description=f"{self.config.description} (Worker {worker_id})",
                config=self.config.config,
                capabilities=self.config.capabilities,
                performance=self.config.performance,
                database=self.config.database,
                monitoring=self.config.monitoring,
                clustering=self.config.clustering
            )
            
            # Create and start server
            server = create_optimized_server(self.server_class, worker_config)
            
            # Run the server
            asyncio.run(server.start())
            
        except Exception as e:
            logging.error(f"Worker {worker_id} failed: {e}")
            raise

# Utility decorators for optimization
def cached_method(ttl: int = 3600):
    """Decorator for caching method results"""
    def decorator(method):
        async def wrapper(self, *args, **kwargs):
            # Create cache key
            cache_key = f"{self.__class__.__name__}:{method.__name__}:{hash((args, tuple(sorted(kwargs.items()))))}"
            
            # Try cache first
            if hasattr(self, 'get_cached'):
                cached_result = await self.get_cached(cache_key)
                if cached_result is not None:
                    return cached_result
            
            # Execute method
            result = await method(self, *args, **kwargs)
            
            # Cache result
            if hasattr(self, 'set_cached'):
                await self.set_cached(cache_key, result, ttl)
            
            return result
        
        return wrapper
    return decorator

def rate_limited(calls_per_second: float):
    """Rate limiting decorator"""
    min_interval = 1.0 / calls_per_second
    last_called = [0.0]
    
    def decorator(method):
        async def wrapper(*args, **kwargs):
            elapsed = time.time() - last_called[0]
            left_to_wait = min_interval - elapsed
            if left_to_wait > 0:
                await asyncio.sleep(left_to_wait)
            result = await method(*args, **kwargs)
            last_called[0] = time.time()
            return result
        return wrapper
    return decorator

def cpu_intensive(use_process_pool: bool = True):
    """Decorator for CPU-intensive methods"""
    def decorator(method):
        async def wrapper(self, *args, **kwargs):
            if use_process_pool and hasattr(self, 'process_cpu_intensive_task'):
                return await self.process_cpu_intensive_task(method, self, *args, **kwargs)
            elif hasattr(self, 'process_io_intensive_task'):
                return await self.process_io_intensive_task(method, self, *args, **kwargs)
            else:
                return await method(self, *args, **kwargs)
        return wrapper
    return decorator

File: mcp_servers/src/core/optimized-base-server.ts
/**
 * Optimized Base MCP Server
 * High-performance implementation with all optimization features enabled
 */

import { BaseMCPServer, MCPServerOptions } from './base-server';
import { PerformanceOptimizer, PerformanceConfig } from './performance-optimizer';
import { ConnectionPoolManager, ConnectionPoolConfig } from './connection-pool';
import { LoadBalancer, createLoadBalancer } from './load-balancer';
import { PerformanceMonitor, createPerformanceMonitor, MonitoringConfig } from './performance-monitor';
import { createServerLogger } from './logger';
import { Logger } from 'pino';
import cluster from 'cluster';
import os from 'os';

export interface OptimizedServerConfig extends MCPServerOptions {
  performance?: Partial<PerformanceConfig>;
  connectionPool?: Partial<ConnectionPoolConfig>;
  monitoring?: Partial<MonitoringConfig>;
  clustering?: {
    enabled: boolean;
    workers?: number;
    gracefulShutdownTimeout?: number;
  };
}

export abstract class OptimizedMCPServer extends BaseMCPServer {
  private performanceOptimizer: PerformanceOptimizer;
  private connectionPool?: ConnectionPoolManager;
  private loadBalancer?: LoadBalancer;
  private performanceMonitor: PerformanceMonitor;
  private clusterId: string;

  constructor(config: OptimizedServerConfig) {
    super(config);
    
    this.clusterId = cluster.worker?.id?.toString() || 'master';
    this.logger = createServerLogger(`${config.name}-${this.clusterId}`);
    
    // Initialize performance optimizer
    this.performanceOptimizer = new PerformanceOptimizer(config.performance, this.logger);
    
    // Initialize connection pool if database config provided
    if (config.connectionPool) {
      this.connectionPool = new ConnectionPoolManager(
        config.connectionPool as ConnectionPoolConfig,
        this.logger
      );
    }

    // Initialize performance monitor
    this.performanceMonitor = createPerformanceMonitor(config.monitoring, this.logger);
    
    // Initialize load balancer for multi-instance deployments
    if (cluster.isWorker) {
      this.loadBalancer = createLoadBalancer('resource-based', {}, this.logger);
    }

    this.setupOptimizations();
    this.setupMonitoring();
  }

  private setupOptimizations(): void {
    // Set up performance optimizer event handlers
    this.performanceOptimizer.on('alert', (alert) => {
      this.logger.warn({ alert }, 'Performance alert triggered');
      this.emit('performance-alert', alert);
    });

    this.performanceOptimizer.on('metrics', (metrics) => {
      // Record metrics in performance monitor
      this.performanceMonitor.recordMetrics(
        this.clusterId,
        metrics,
        this.connectionPool ? this.connectionPool.getAllMetrics() : {}
      );
    });

    // Connection pool event handlers
    if (this.connectionPool) {
      this.connectionPool.on('error', (error) => {
        this.logger.error({ error }, 'Connection pool error');
        this.performanceOptimizer.trackError();
      });

      this.connectionPool.on('db-metrics', (metrics) => {
        this.logger.debug({ metrics }, 'Database metrics updated');
      });
    }

    // Load balancer event handlers
    if (this.loadBalancer) {
      this.loadBalancer.on('server-failed', (server) => {
        this.logger.warn({ serverId: server.id }, 'Load balancer detected server failure');
      });

      this.loadBalancer.on('scale-up-needed', (data) => {
        this.logger.info({ data }, 'Auto-scaling: scale up needed');
        this.emit('scale-up-needed', data);
      });

      this.loadBalancer.on('scale-down-needed', (data) => {
        this.logger.info({ data }, 'Auto-scaling: scale down possible');
        this.emit('scale-down-needed', data);
      });
    }
  }

  private setupMonitoring(): void {
    // Set up performance monitor event handlers
    this.performanceMonitor.on('alert', (alert) => {
      this.logger.warn({ alert }, 'Monitoring alert triggered');
      this.emit('monitoring-alert', alert);
    });

    this.performanceMonitor.on('metrics-recorded', (snapshot) => {
      this.logger.debug('Performance metrics recorded');
    });
  }

  // Enhanced request handling with performance optimizations
  protected async handleOptimizedRequest<T>(
    requestName: string,
    handler: () => Promise<T>,
    cacheKey?: string,
    cacheTTL?: number
  ): Promise<T> {
    const startTime = Date.now();
    this.performanceOptimizer.trackRequest();

    try {
      // Try cache first if cache key provided
      if (cacheKey) {
        const cachedResult = await this.performanceOptimizer.get<T>(cacheKey);
        if (cachedResult !== null) {
          this.performanceOptimizer.trackResponseTime(Date.now() - startTime);
          return cachedResult;
        }
      }

      // Process request asynchronously
      const result = await this.performanceOptimizer.processAsync(handler);

      // Cache result if cache key provided
      if (cacheKey) {
        await this.performanceOptimizer.set(cacheKey, result, cacheTTL);
      }

      this.performanceOptimizer.trackResponseTime(Date.now() - startTime);
      this.logger.info({ requestName, duration: Date.now() - startTime }, 'Request processed');

      return result;
    } catch (error) {
      this.performanceOptimizer.trackError();
      this.performanceOptimizer.trackResponseTime(Date.now() - startTime);
      this.logger.error({ error, requestName, duration: Date.now() - startTime }, 'Request failed');
      throw error;
    }
  }

  // CPU-intensive task processing with worker threads
  protected async processCPUIntensiveTask<T>(data: any): Promise<T> {
    return this.performanceOptimizer.processWithWorker<T>(data);
  }

  // Database operations with connection pooling
  protected async queryDatabase<T>(
    sql: string,
    params?: any[],
    dbType: 'postgres' | 'mysql' = 'postgres'
  ): Promise<T> {
    if (!this.connectionPool) {
      throw new Error('Connection pool not configured');
    }

    if (dbType === 'postgres') {
      return this.connectionPool.queryPostgres<T>(sql, params);
    } else {
      return this.connectionPool.queryMySQL<T>(sql, params);
    }
  }

  protected async transactionDatabase<T>(
    callback: (client: any) => Promise<T>,
    dbType: 'postgres' | 'mysql' = 'postgres'
  ): Promise<T> {
    if (!this.connectionPool) {
      throw new Error('Connection pool not configured');
    }

    if (dbType === 'postgres') {
      return this.connectionPool.transactionPostgres(callback);
    } else {
      return this.connectionPool.transactionMySQL(callback);
    }
  }

  // HTTP requests with connection pooling
  protected getOptimizedHttpAgent(): import('http').Agent {
    if (!this.connectionPool) {
      throw new Error('Connection pool not configured');
    }
    return this.connectionPool.getHttpAgent();
  }

  protected getOptimizedHttpsAgent(): import('https').Agent {
    if (!this.connectionPool) {
      throw new Error('Connection pool not configured');
    }
    return this.connectionPool.getHttpsAgent();
  }

  // Cache operations
  protected async getCached<T>(key: string): Promise<T | null> {
    return this.performanceOptimizer.get<T>(key);
  }

  protected async setCached<T>(key: string, value: T, ttl?: number): Promise<void> {
    return this.performanceOptimizer.set(key, value, ttl);
  }

  protected async invalidateCache(key: string): Promise<void> {
    return this.performanceOptimizer.invalidate(key);
  }

  protected async invalidateCachePattern(pattern: string): Promise<void> {
    return this.performanceOptimizer.invalidatePattern(pattern);
  }

  // Performance metrics
  getPerformanceMetrics() {
    return this.performanceOptimizer.getMetrics();
  }

  getConnectionMetrics() {
    return this.connectionPool ? this.connectionPool.getAllMetrics() : {};
  }

  getLoadBalancerStats() {
    return this.loadBalancer ? this.loadBalancer.getStatistics() : null;
  }

  getCurrentMonitoringSnapshot() {
    return this.performanceMonitor.getCurrentMetrics();
  }

  // Health checks
  async getDetailedHealth() {
    const baseHealth = await this.getHealth();
    
    return {
      ...baseHealth,
      performance: this.performanceOptimizer.getMetrics(),
      connectionPools: this.connectionPool ? await this.connectionPool.healthCheck() : {},
      loadBalancer: this.loadBalancer ? this.loadBalancer.getStatistics() : null,
      monitoring: {
        activeAlerts: this.performanceMonitor.getActiveAlertCount(),
        uptime: Math.floor(process.uptime())
      }
    };
  }

  // Enhanced cleanup
  protected async cleanup(): Promise<void> {
    this.logger.info('Starting optimized server cleanup');

    const cleanupPromises: Promise<void>[] = [];

    // Cleanup performance optimizer
    cleanupPromises.push(this.performanceOptimizer.cleanup());

    // Cleanup connection pool
    if (this.connectionPool) {
      cleanupPromises.push(this.connectionPool.close());
    }

    // Cleanup load balancer
    if (this.loadBalancer) {
      this.loadBalancer.destroy();
    }

    // Cleanup performance monitor
    cleanupPromises.push(this.performanceMonitor.cleanup());

    await Promise.all(cleanupPromises);

    this.logger.info('Optimized server cleanup completed');
  }
}

// Factory function for creating optimized servers
export function createOptimizedServer<T extends OptimizedMCPServer>(
  ServerClass: new (config: OptimizedServerConfig) => T,
  config: OptimizedServerConfig
): T {
  // Apply default optimizations for AMD Ryzen 7 7800X3D
  const optimizedConfig: OptimizedServerConfig = {
    ...config,
    performance: {
      enableWorkerThreads: true,
      workerPoolSize: Math.min(8, os.cpus().length - 2),
      threadPoolSize: os.cpus().length,
      enableAsyncBatching: true,
      maxMemoryUsage: 28 * 1024, // 28GB for 32GB system
      gcThreshold: 75,
      enableMemoryProfiling: true,
      lruCacheSize: 10000,
      cacheTTL: 3600,
      enableRedisCache: true,
      dbConnectionPoolSize: 32,
      httpConnectionPoolSize: 64,
      keepAliveTimeout: 30000,
      enableMetrics: true,
      metricsInterval: 1000,
      alertThresholds: {
        cpuUsage: 85,
        memoryUsage: 90,
        responseTime: 5000
      },
      ...config.performance
    },
    connectionPool: {
      http: {
        maxSockets: 64,
        maxFreeSockets: 16,
        timeout: 30000,
        keepAlive: true,
        keepAliveMsecs: 30000,
        freeSocketTimeout: 15000,
        socketActiveTTL: 300000
      },
      ...config.connectionPool
    },
    monitoring: {
      dashboard: {
        enabled: true,
        port: 3001,
        host: '0.0.0.0',
        updateInterval: 1000
      },
      prometheus: {
        enabled: true,
        port: 9090,
        endpoint: '/metrics'
      },
      alerts: {
        enabled: true,
        thresholds: {
          cpuUsage: 80,
          memoryUsage: 85,
          responseTime: 5000,
          errorRate: 5,
          connectionPoolUtilization: 90
        }
      },
      retention: {
        metricsRetentionHours: 24,
        samplingInterval: 5000
      },
      ...config.monitoring
    },
    clustering: {
      enabled: true,
      workers: os.cpus().length,
      gracefulShutdownTimeout: 30000,
      ...config.clustering
    }
  };

  return new ServerClass(optimizedConfig);
}

// Cluster management for multi-process deployment
export function startClusteredServer<T extends OptimizedMCPServer>(
  ServerClass: new (config: OptimizedServerConfig) => T,
  config: OptimizedServerConfig
): void {
  const numWorkers = config.clustering?.workers || os.cpus().length;

  if (cluster.isPrimary) {
    console.log(`Starting ${numWorkers} workers for ${config.name}`);

    // Fork workers
    for (let i = 0; i < numWorkers; i++) {
      cluster.fork();
    }

    cluster.on('exit', (worker, code, signal) => {
      console.log(`Worker ${worker.process.pid} died. Restarting...`);
      cluster.fork();
    });

    // Graceful shutdown
    process.on('SIGTERM', () => {
      console.log('Primary received SIGTERM, shutting down workers...');
      
      Object.values(cluster.workers || {}).forEach(worker => {
        if (worker) {
          worker.kill('SIGTERM');
        }
      });

      setTimeout(() => {
        process.exit(0);
      }, config.clustering?.gracefulShutdownTimeout || 30000);
    });

  } else {
    // Worker process
    const server = createOptimizedServer(ServerClass, config);
    
    server.start().catch(error => {
      console.error('Failed to start worker server:', error);
      process.exit(1);
    });

    // Graceful shutdown for worker
    process.on('SIGTERM', async () => {
      console.log(`Worker ${process.pid} received SIGTERM, shutting down gracefully...`);
      await server.gracefulShutdown();
    });
  }
}